ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ExplicitDisclosureIT.scala

```diff
@@ -0,0 +1,421 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.testtool.suites.v1_dev
+
+import com.daml.error.definitions.LedgerApiErrors
+import com.daml.ledger.api.testtool.infrastructure.Allocation._
+import com.daml.ledger.api.testtool.infrastructure.Assertions._
+import com.daml.ledger.api.testtool.infrastructure.LedgerTestSuite
+import com.daml.ledger.api.testtool.infrastructure.TransactionHelpers.createdEvents
+import com.daml.ledger.api.testtool.infrastructure.participant.ParticipantTestContext
+import com.daml.ledger.api.v1.command_service.SubmitAndWaitRequest
+import com.daml.ledger.api.v1.commands.{Command, DisclosedContract, ExerciseByKeyCommand}
+import com.daml.ledger.api.v1.event.CreatedEvent
+import com.daml.ledger.api.v1.transaction_service.GetTransactionsRequest
+import com.daml.ledger.api.v1.value.{Record, RecordField, Value}
+import com.daml.ledger.client.binding
+import com.daml.ledger.client.binding.Primitive
+import com.daml.ledger.test.model.Test._
+import com.daml.lf.value.Value.ContractId
+import scalaz.syntax.tag._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+// TODO ED: Enable in Sandbox-on-X conformance test-suite
+final class ExplicitDisclosureIT extends LedgerTestSuite {
+  import ExplicitDisclosureIT._
+
+  test(
+    "EDCorrectDisclosure",
+    "Submission is successful if the correct disclosure is provided",
+    allocate(Parties(2)),
+    enabled = _.explicitDisclosure,
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    for {
+      testContext <- initializeTest(ledger, owner, delegate)
+
+      // Exercise a choice on the Delegation that fetches the Delegated contract
+      // Fails because the submitter doesn't see the contract being fetched
+      exerciseFetchError <- testContext
+        .exerciseFetchDelegated()
+        .mustFail("the submitter does not see the contract")
+
+      // Exercise the same choice, this time using correct explicit disclosure
+      _ <- testContext.exerciseFetchDelegated(testContext.disclosedContract)
+
+      // Create an extra disclosed contract
+      extraKey = ledger.nextKeyId()
+      _ <- ledger.create(owner, Delegated(owner, extraKey))
+      delegatedTxs <- ledger.flatTransactionsByTemplateId(Delegated.id, owner)
+      extraDelegatedEvent = createdEvents(delegatedTxs(1)).head
+      extraDisclosedContract = createEventToDisclosedContract(extraDelegatedEvent)
+
+      // Exercise the same choice, this time with superfluous disclosure
+      _ <- testContext.exerciseFetchDelegated(testContext.disclosedContract, extraDisclosedContract)
+    } yield {
+      assertGrpcError(
+        exerciseFetchError,
+        LedgerApiErrors.ConsistencyErrors.ContractNotFound,
+        None,
+        checkDefiniteAnswerMetadata = true,
+      )
+    }
+  })
+
+  test(
+    "EDExerciseByKeyDisclosedContract",
+    "A disclosed contract can be exercised by key by non-stakeholder if authorized",
```

Comment:
the description is slightly confusing. What we’re interested in is not the actor of the exercise but the set of reading parties (actAs union readAs).